### ISSUE TITLE
Invalid passed (or failed) test count for launch

### ISSUE THREAD

**Issue by ekisielinski** [reportportal/reportportal#735]
I think I have similar issue to: #724 

We use Bamboo + RP NUnit plugin to run tests.
Bamboo shows different total test count than ReportPortal shows:

![reportportal](https://user-images.githubusercontent.com/22278763/67570167-08867b00-f731-11e9-8c2a-32685410a9b2.png)

Launch says there was 20 tests passed, but when I click it - it shows only 6.
Bamboo shows less tests executed (correct number) and RP show incorrect (higher number).
-----
Current version: Analysis Service: 4.3.0; API Service: 4.3.7; Index Service: 4.2.0; Jira Service: 4.3.0; Authorization Service: 4.2.0; Service UI: 4.3.7;

**Comment by e-olga** [comment]
hello @ekisielinski,
When you click the statistics link (e.g., from the Launches page), you are redirected to STEP view of items with "type"="STEP".
In case your structure of test looks like Suite>Test>Step, all the steps will be displayed on the STEP view, and the statistics you clicked will be equal to items displayed on the view.
In case the structure of your test looks like Suite>Test or Suite>Test>Test, etc the statistics on the page would be different in compare with statistics link.
So that, to have the expected number of tests when you click the statistics link, you should have the appropriate structure of all the tests.
P.s., you are right, that your issue is similar to the #724)

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko might help here

**Comment by nvborisenko** [comment]
NUnit reports tests as Suite - Suite - Step.

**Comment by ekisielinski** [comment]
@e-olga: Sorry but I don't understand what you have tried to say :) I don't see any "statistics link".

Some tests (methods) are visible in RP, but they don't have any logs inside.

Here is (I think) suite with nothing inside, but it shows passed = 1

![errrrrrrr](https://user-images.githubusercontent.com/22278763/67752779-e728d600-fa33-11e9-9983-ccbeec419b83.png)

**Comment by e-olga** [comment]
@ekisielinski, Regarding the statistics link, https://i.imgur.com/cHm1VVJ.png.
In case the test does not have item with "type"="STEP", RP in any case calculates the statistics, but you will not see any item on STEP view.

**Comment by evjlobanova** [comment]
@ekisielinski  Have you resolved your issue? Please let us know if we can close a ticket.

**Comment by nvborisenko** [comment]
The problem is empty suites, old problem. UI shows this empty item incorrectly. See screenshots above.

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko v5 api will allow to delete them before launch finish.
will it help?

**Comment by nvborisenko** [comment]
It is good that I can see empty suite in RP, this is my testing result.

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko but you need zero stats for it, right?

**Comment by nvborisenko** [comment]
Yes, zero, if it helps and doesn't break your model. I watch at `suites/tests/steps` as at `folders/files`. I may have empty folder, and it is strange for me that empty folder has 1 file inside.

**Comment by ekisielinski** [comment]
@evjlobanova Not solved. Why it shows empty suites? Idk. If suite is empty why it tells something passed? Idk.

**Comment by DzmitryHumianiuk** [comment]
@ekisielinski specific of the test runner + specifics of RP stats counting.

Runner triggers events for start and stop even for empty suite.
RP assumes element with no child as final tree branch. And set stats as 1, and passed since it’s passed.

Team discuss options for fix

**Comment by ekisielinski** [comment]
@DzmitryHumianiuk OK, I understand the problem with "invalid" statistics, but I still don't know why I have empty suite.
I was looking in our tests and that suite is not empty. It should not be even selected by NUnit runner due to category filters.

I'll look at this problem again soon.

**Comment by nvborisenko** [comment]
@ekisielinski does default `test-results.xml` have this empty suite? If yes - probably this is issue of nunit engine, if no - we should resolve the issue on RP agent side.

**Comment by ekisielinski** [comment]
@nvborisenko: You are right. NUnit xml result file contains empty suites.

NUnit is a garbage...

**Comment by nvborisenko** [comment]
@ekisielinski you didn't use others.

Actually I cannot reproduce empty suites in report. When I execute tests by category filter and tests in suite don't fit this filter, I don't see empty suite in report.

I propose to create an issue for nunit. If this issue will be fixed soon - good. In others cases possible solutions are:
- fix on server side to correctly accept empty suites
- fix on agent side to delete empty suites

**Comment by DzmitryHumianiuk** [comment]
We plan fix on server side.
Actually possibility to send ZeroStat object

**Comment by nvborisenko** [comment]
@DzmitryHumianiuk what is ZeroStat object? The agent knows about empty suite at he the end of suite's execution.

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko in v5 we allow to send update rq for testitems.
Triggered event for empty suite at the end of it can send testItem update with zero statistics.

Zeros for Total, Passed, Failed, Skipped, etc.

**Comment by nvborisenko** [comment]
@DzmitryHumianiuk does it mean that agent should send additional request to update "empty" suite? If yes, agent can delete "empty" suite. This way is compatible with v4 and v5 server.

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko 
It can be opted:

1) send delete request for empty suite, if it’s not important to have it empty in RP

2) before finish RQ send RQ to check child count. If empty - send finishRQ with zero stats. Or if you know it from runner info that suite is empty.

First compatible with v4/5
Second will be added on server side

**Comment by nvborisenko** [comment]
Runner/agent knows that suite is empty.

**Comment by DzmitryHumianiuk** [comment]
@nvborisenko then second option is for us.
will be in 5.1

**Comment by ekisielinski** [comment]
Displaying 0 stats or removing empty suites will be sufficient for me :)

**Comment by nvborisenko** [comment]
Cannot delete empty suites when launch is in progress state.

`{"error_code":4063,"message":"Unable to perform operation for non-finished launch. Unable to delete test item ['5deabacc69515b000113477f'] under launch ['5deabacc69515b0001134778'] with 'In progress' state"})`

**Comment by nvborisenko** [comment]
Given that:
 - agent is not able to delete empty suites in run-time
 - agent doesn't want to delete empty suites after tests execution because of RP is directed to be real-time reporting system.

My conclusion is the following: this issue should be resolved on server side to properly handle empty suites.. if NUnit team will not fix the issue in near future...

**Comment by evjlobanova** [comment]
A possibility to delete test items in run-time has been added to the Backlog
